### PR TITLE
SLIME deactivation: Only when really activated/loaded, prompt once or none

### PR DESCRIPTION
**Pull request by OlCe2**
Deactivating SLIME was always done if `slime-lisp-mode-hook' was not in
`lisp-mode-hook' or the process was not interactive, and it was done by calling
`slime-mode'... which triggers automatic loading of SLIME, defeating the
already imperfect deactivation (which, for some SLIME's contrib, can't be done
easily).

Fix this by removing `slime-lisp-mode-hook' only if present on the hook, and
deactivating `slime-mode' in buffers only if really activated
(which means SLIME has been loaded).

Prompt the user only once for all occurences, and none if there are none.

### CODE REVIEW

**Comment by joaotavora** [comment]
Uff.  I must say I don't understand the problem.  I'm not saying there isn't any. But can you walk me through the your scenario slowly, in clear steps, in a way that highlights the problem in the current code?

Anyway, I think I understand vaguely.  You're saying that `slime-lisp-mode-hook` might be there, but `slime.el` not actually loaded?  And then `(slime-mode -1)` would end up triggering the autoload, is that it?

Nevertheless, please provide an idiot-proof recipe that I can follow.

**Comment by joaotavora** [comment]
If the problem is indeed what I think it is, then I think it's a question of checking where the variable `slime-mode` is bound, or some other indication of thje `slime.el` library actually being loaded.

Anyway, please bear in mind that this code is meant for letting people try SLY quickly.  A preferred installation of SLY shouldn't coexist with one for SLIME.

**Comment by OlCe2** [comment]
> Anyway, I think I understand vaguely. You're saying that slime-lisp-mode-hook
might be there, but slime.el not actually loaded? And then (slime-mode -1)
would end up triggering the autoload, is that it?
> Nevertheless, please provide an idiot-proof recipe that I can follow.

Sorry, I should have added much more context, because indeed in "normal" cases
(i.e., without customizations or manually loading autoloads) this does not
happen.

Original SLY's code does this:
1. Decide whether SLIME should be deactivated.
2. When SLIME must be deactivated:
a. Remove `slime-lisp-mode-hook` from `slime-lisp-mode-hook`.
b. Try calling `(slime-mode -1)` in each buffer already in Lisp mode.

SLIME's recommended procedure to enable autoloading is to require
`slime-autoloads.el`. But this file, in addition to defining some autoloads,
also pushes `slime-lisp-mode-hook` into `lisp-mode-hook`.

In this case, step 1 above decides that SLIME has to be deactivated because
`slime-lisp-mode-hook` is present in hook variable `lisp-mode-hook`. Then, if
there already exists a buffer in Lisp mode, step 2b triggers SLIME's full load
if not already loaded. So there is a problem here iff there exists such a
buffer that didn't already trigger a load of SLIME.

In the usual scenario of starting Emacs with SLIME's and SLY's autoloads and
immediately activating SLY (e.g., creating a Lisp session), no, because either
the user did not open any Lisp buffer before activating SLY, or he did, but
since `slime-lisp-mode-hook` is in `lisp-mode-hook`, activation of Lisp mode
already caused SLIME to be loaded, and step 2b does not do more harm in this
respect.

However, slight differences to this scenario cause the problem to appear. First
time I saw it was with this sequence:

1. Start Emacs, without SLIME/SLY autoloads.
2. Open a Lisp file.
3. Require SLIME autoloads.
4. Change your mind, and load SLY.

There, there exists a Lisp-mode buffer and SLIME was not loaded before 4, so 4
triggers loading of SLIME as just explained above, which was not intended.

You may argue that the previous scenario is anecdotical, and I concur, but
unfortunately what finally prompted me to do the change is something much more
annoying.

My goal is to allow the possibility of quickly launching SLIME in some Emacs
processes and SLY in some others. To this end, I require autoloads for both
modules to be present. But then, opening a Lisp file without having first
opened a SLIME session or loaded SLY causes SLIME to be loaded (because
`slime-lisp-mode-hook` is in `lisp-mode-hook`). I consider this a problem,
which I fixed by hacking `slime-autoloads.el` to remove the adding of
`slime-lisp-mode-hook` to `lisp-mode-hook`. Without going that far, anybody
could simply remove the hook in one of its init file.

And this is what uncovers the problem in SLY's code: Step 1 decides that SLIME
must be deactivated *even when* `slime-lisp-mode-hook` is not in
`lisp-mode-hook`, and the user is not even prompted in this case! So if I open
a Lisp file and only after that make my mind to use SLY, step 2b will indeed
erroneously load SLIME, which I precisely didn't want.

> If the problem is indeed what I think it is, then I think it's a question of
checking where the variable slime-mode is bound, or some other indication of
thje slime.el library actually being loaded.

The change I'm proposing indeed uses `slime-mode`: First test, for each buffer
in Lisp mode, if this variable is actually bound, and to true, and only then
deactivate slime-mode in the buffer.

While here, I thought it to be much more clear and correct to simply have
separate tests before running step 2a and 2b respectively, instead of the
current umbrella test at step 1. The `proceed-p` part allows to prompt the user
exactly when required, i.e., at first hint of SLIME being auto-loaded or fully
loaded, as long as the Emacs process is connected to a terminal.

> Anyway, please bear in mind that this code is meant for letting people try
SLY quickly. A preferred installation of SLY shouldn't coexist with one for
SLIME.

I think this depends a lot on what you mean exactly by "installation" and
"coexist". If you mean loading both SLIME and SLY in a single Emacs process,
then I agree. But if you mean dismissing a mere installation of files for both,
or even the requiring of both autoloads, which lets the user choose between
SLIME and SLY up to the last moment, then I don't, especially considering that
the change I propose is small, achieves exactly that (with the companion tweak
to remove `slime-lisp-mode-hook') and doesn't change the previous behavior in the usual
scenarios (such as the one I described at start).

**Comment by monnier** [comment]
> I think this depends a lot on what you mean exactly by "installation" and
> "coexist". If you mean loading both SLIME and SLY in a single Emacs process,
> then I agree. But if you mean dismissing a mere installation of files for both,
> or even the requiring of both autoloads, which lets the user choose between
> SLIME and SLY up to the last moment, then I don't,

FWIW, I fully agree.  It would make a lot of sense for a shared server
to have both SLY and SLIME installed as systemwide ELPA packages, so by
default both autoloads would be setup when launching an Emacs session.

**Comment by OlCe2** [comment]
> FWIW, I fully agree. It would make a lot of sense for a shared server to have both SLY and SLIME installed as systemwide ELPA packages, so by default both autoloads would be setup when launching an Emacs session.

Yes, and this would be a step in that direction.

There would be two steps missing though:
1. Remove addition of `slime-lisp-mode-hook` into `lisp-mode-hook` in SLIME.
2. Conversely, remove addition of `sly-editing-mode` into `lisp-mode-hook` in SLY.

These ones could be controversial because, by contrast with the change proposed in this PR, they actually change the default behavior of each package (i.e., opening a Lisp file normally triggers automatic loading of SLIME/SLY with proper fontification, completion, etc.).

Also, it is possible to obtain the same result by explicitly requiring the autoloads of both projects and then reverting the changes in `lisp-mode-hook` in some Emacs init file.

But there is no such workaround for the problem fixed by this PR.

**Comment by joaotavora** [comment]
> There, there exists a Lisp-mode buffer and SLIME was not loaded before 4, so 4 triggers loading of SLIME as just explained above, which was not intended.

Correct, then I think I understood correctly and this should indeed be fixed.  So I'm quite receptive to your change.

> The change I'm proposing indeed uses slime-mode: First test, for each buffer in Lisp mode, if this variable is actually bound, and to true, and only then deactivate slime-mode in the buffer.

Right. But you shouldn't have to test for each buffer.  You can leverage `(featurep 'slime)` for example to know if the autoloads have kicked in, for example. That doesn't look too brittle.  Previously I suggested checking if `slime-mode` is `fbound`, but it's still fbound if there's an autoload cookie on it.

> I think this depends a lot on what you mean exactly by "installation" and "coexist". If you mean loading both SLIME and SLY in a single Emacs process, then I agree. But if you mean dismissing a mere installation of files for both, o

I did mean the former, but with a caveat.  You can of course have autolaods set up for both SLIME and SLY, but as I think you understand , you can't work with both _at the same time_, because they will try to take-over your `lisp-mode` buffers in conflicting ways, such as keybindings.

Here's the caveat:  SLY has code to detect SLIME, but the inverse isn't true. So if you load SLIME then SLY, you are given options to (try to) preserve consistency, but if you do it the other way around, you don't.  So beware.  

As far as I know Emacs doesn't have a generic mechanism to automatically handle a handover of ownership, so there's a footgun here.

**Comment by joaotavora** [comment]
> Remove addition of slime-lisp-mode-hook into lisp-mode-hook in SLIME.
> These ones could be controversial because, by contrast with the change proposed in this PR, they actually change the default behavior of each package

Yes quite "controversial". Unless you came up with an alternative. you'd be breaking year, decades' worth of SLIME and SLY use cases for people everywhere.

**Comment by joaotavora** [comment]
I think this one-line patch should address your use cases:
```lisp
diff --git a/sly.el b/sly.el
index 37567c31..36dbceb8 100644
--- a/sly.el
+++ b/sly.el
@@ -7464,7 +7464,7 @@ and replace `sly-editing-mode' with `slime-lisp-mode-hook'.")))
     (with-current-buffer buffer
       (when (eq major-mode 'lisp-mode)
         (sly-editing-mode 1)
-        (ignore-errors (funcall 'slime-mode -1))))))
+        (ignore-errors (and (featurep 'slime) (funcall 'slime-mode -1)))))))
  (t
   (warn
    "`sly.el' loaded OK. To use SLY, customize `lisp-mode-hook' and remove `slime-lisp-mode-hook'.")))
```

**Comment by OlCe2** [comment]
> Right. But you shouldn't have to test for each buffer. You can leverage `(featurep 'slime)` for example to know if the autoloads have kicked in, for example. That doesn't look too brittle. Previously I suggested checking if `slime-mode` is `fbound`, but it's still fbound if there's an autoload cookie on it.

I still prefer the proposed change because it is "minimal": It only tries to deactivate slime-mode in buffers where it is activated. I expect that in most cases, it will be activated in all Lisp buffers, but why ruling out corner cases, when you can have them taken into account easily?
 
> I did mean the former, but with a caveat. You can of course have autolaods set up for both SLIME and SLY, but as I think you understand , you can't work with both _at the same time_, because they will try to take-over your `lisp-mode` buffers in conflicting ways, such as keybindings.

Sure, I understand that, that's why I agree with the former.

> Here's the caveat: SLY has code to detect SLIME, but the inverse isn't true. So if you load SLIME then SLY, you are given options to (try to) preserve consistency, but if you do it the other way around, you don't. So beware.

That's true, but with some SLIME contribs, I fear that SLY is not doing enough to revert SLIME's changes. So I much prefer the alternative where neither package makes too deep modifications from their autoloads in the first place, because then there's nothing complicated to revert.
 
> As far as I know Emacs doesn't have a generic mechanism to automatically handle a handover of ownership, so there's a footgun here.

There is apparently some support in Emacs for unloading some package completely (see, e.g., Loading/Unloading in the [Emacs Lisp manual](http://www.gnu.org/software/emacs/manual/html_mono/elisp.html#Unloading)), but I've not tested it. But this would be just a step in this direction, which could be achieved through different means.

In any case, I don't think that we can reasonably expect one project to correctly write code to deactivate another project without support from the latter. That's another reason why I think we should prefer the approach of minimum modifications at autoload.

**Comment by joaotavora** [comment]
@OlCe2  please test

**Comment by joaotavora** [comment]
I>  still prefer the proposed change because it is "minimal": 

Really?  More minimal than my ~20-char change?

> but why ruling out corner cases, when you can have them taken into account easily?

What corner cases?  Unless I'm missing something, it seems it's trying to protect against 2+2 ever being 5.

> So I much prefer the alternative where neither package makes too deep modifications from their autoloads in the first place, because then there's nothing complicated to revert.

Maybe I would like this world too, but that is not the real world :-).  Users want to load the SLIME/SLY and have them automatically setup `slime-mode` and/or `sly-mode` for all the buffers that are already in Emacs's lisp-mode.  

> There is apparently some support in Emacs for unloading some package completely (see, e.g., Loading/Unloading in the Emacs Lisp manual), but I've not tested it. But this would be just a step in this direction, which could be achieved through different means.

Yes, I know this "support".  It's a best effort thing,it doesn't magically revert the memory effects of loading a package, that's not how Elisp works.

**Comment by OlCe2** [comment]
> Really? More minimal than my ~20-char change?

I'm not talking about the number of characters you change, but the amount of Emacs state you are browsing and acting upon. Such as:
1. Calling `(slime-mode -1)` in each Lisp buffer even when slime-mode is not activated (what's the point? plus you risk creating problems if slime-mode at some point has a bug in this case).
2. Testing `(featurep 'slime)` in all buffers. This makes sense for `slime-mode`, which is buffer local, but not for features.
 
> What corner cases? Unless I'm missing something, it seems it's trying to protect against 2+2 ever being 5.

Corner cases such as manual deactivation of slime-mode in some buffer because it is causing problems. Then again, you'll force deactivation of something that is already deactivated.
 
> Maybe I would like this world too, but that is not the real world :-). Users want to load the SLIME/SLY and have them automatically setup `slime-mode` and/or `sly-mode` for all the buffers that are already in Emacs's lisp-mode.

I was just here talking about the proposed change. And yes, this is the real world: I can achieve the goal I stated, as I already described. I understand that not automatically setting up `slime-mode` or `sly-mode` is controversial, I'm not pushing for it now, I just think it could be great that both projects provide a mean to revert such changes (and really, this is not hard). And as said, there are workarounds, so I can live with that.

**Comment by OlCe2** [comment]
> @OlCe2 please test

The scenarios I listed indeed now work with your change. But please see my previous message.

**Comment by joaotavora** [comment]
> Then again, you'll force deactivation of something that is already deactivated.

But what's the practical problem of that?  `slime-mode -1` is idempotent by design (and by requirement of minor modes).

**Comment by joaotavora** [comment]
> (what's the point? plus you risk creating problems if slime-mode at some point has a bug in this case).

If `slime-mode` somehow develops a bug, then why is it more likely to cause problems when called n times than when called just once? Even accepting your somewhat far-fetched extrapolations, this is totally unclear to me.

> both projects provide a mean to revert such changes (and really, this is not hard)

If you want to provide a PR for SLY to "revert" its changes to Emacs state, by means of the "best effort" thing mentioned above, I am open to reviewing it.

**Comment by OlCe2** [comment]
> But what's the practical problem of that? `slime-mode -1` is idempotent by design (and by requirement of minor modes).

Currently, none. It's just that doing this is exercizing more code than necessary, and thus has more chance to trigger potential future bug.

Nobody seems to have seen the problem I described here between the autoload and clean-up of SLIME. It would have been avoided from the start if `slime-mode -1` had been called only if necessary. The unlikely does indeed happen sometimes. It's just the same reasoning applied again. And, yes, idempotence is a requirement of minor modes... which is not automatically enforced. Are you betting this will always be the case? When I have easy means not to depend on that, I avoid relying on it.

**Comment by OlCe2** [comment]
> If `slime-mode` somehow develops a bug, then why is it more likely to cause problems when called n times than when called just once? Even accepting your somewhat far-fetched extrapolations, this is totally unclear to me.

It's not a problem of being called N times, but rather the contexts of the calls. In once case you call `slime-mode -1` only when `slime-mode` is `t`, in the other case you call `slime-mode -1` with different and varying values of `slime-mode`.

**Comment by OlCe2** [comment]
> If you want to provide a PR for SLY to "revert" its changes to Emacs state, by means of the "best effort" thing mentioned above, I am open to reviewing it.

Thanks, probably at some point. However all contribs must be reviewed thoroughly for the changes they make and tested, and I don't have enough time for this now.

**Comment by joaotavora** [comment]
> Currently, none. It's just that doing this is exercising more code than necessary, and thus has more chance to trigger potential future bug.

Calling `slime-mode -1` more than once exercises more code than calling it once?  Extremely unlikely, except in the most far-fetched scenarios.  Look if we're already going to rely on `slime-mode -1` to do this correctly, why should we also  distrust `slime-mode -1` to be misbehaving in insanely contorned ways?  Why should we spends lines of code and complexity in plugging  imaginary hole while leaving the main door open?

> Nobody seems to have seen the problem I described here between the autoload and clean-up of SLIME

That's because it's completely artificial.  People normally don't load manually load autoloads manually in their session, because autoloads are really meant to be loaded _automatically_.

>  The unlikely does indeed happen sometimes. 

Right, but I still believe in science. :scientist:

**Comment by OlCe2** [comment]
> Calling `slime-mode -1` more than once exercises more code than calling it once? Extremely unlikely, except in the most far-fetched scenarios. Look if we're already going to rely on `slime-mode -1` to do this correctly, why should we also distrust `slime-mode -1` to be misbehaving in insanely contorned ways?

You don't understand. Please re-read my comment above the context, and how it changes between the calls in both cases (summary in other words: in one case, idempotence is relied upon, in the other case, not at all).

> Why should we spends lines of code and complexity in plugging imaginary hole while leaving the main door open?

Which complexity? The one of testing if we are in certain state before performing action on that state? Is this really more complex than having an umbrella test which proved problematic in this very case? And we are talking about a few lines of codes, does this really cost you a lot?

Which main door? I don't want to leave any, as you must have understood. But I don't think there is any. This is what sounds imaginary.

There are potential problematic scenarios which doing things minimally avoids completely. Some of them are very close to materialize, they just require a flaw in idempotence of slime-mode. And this is so easy to avoid. Yet you prefere to ignore that. You're the one leaving the door (arguably small, but still) open.

> > Nobody seems to have seen the problem I described here between the autoload and clean-up of SLIME
> 
> That's because it's completely artificial. People normally don't load manually load autoloads manually in their session, because autoloads are really meant to be loaded _automatically_.

Call it artificial all you want, it corresponds to some real use case, and may very well become the standard in the future.

>> The unlikely does indeed happen sometimes.
> Right, but I still believe in science.

If so, then perhaps you should start by realizing that your last statement is completely unrelated to the one you quoted immediately above. ;-)

**Comment by joaotavora** [comment]
> You don't understand. Please re-read my comment above the context, and how it changes between the calls in both cases (summary in other words: in one case, idempotence is relied upon, in the other case, not at all).

Indeed, I don't understand.  Calling `(slime-mode -1)` for every lisp buffer is fine if that function exists.  I guess I could check if `slime-mode` is non-nil before doing it, but that's the job of `slime-mode` anyway.  So it just sounds useless.

> Which complexity? 

All the lines of code and code paths that you proposed to add.

> Call it artificial all you want, it corresponds to some real use case, and may very well become the standard in the future.

That it _can_ conceivably happen is the reason I patched it, right?  I solved the problem you were having, did I not?  I really really doubt loading autoloads manually will become "standard" in the future.  Let me know when it does.

> If so, then perhaps you should start by realizing that your last statement is completely unrelated to the one you quoted immediately above. ;-)

I could also have said: "I don't believe in witches", but there's no funny emoji for that.

**Comment by OlCe2** [comment]
> I solved the problem you were having, did I not?

Yes, thanks. I expected more than that, but that's indeed much better than nothing.

**Comment by joaotavora** [comment]
It was the only real problem I could see.  Sorry I didn't meet your expectations. Thanks for using SLY. :+1:
